### DISCUSSION TITLE
Lesson 9: Test for the Interaction.s.sol | Critique and suggest : )

### DISCUSSION THREAD

**Discussion by sharonphiliplima** [Cyfrin/foundry-full-course-cu#486] [Q&A]
I am stuck on the problem:
There are three contracts -
`import {CreateSubscription, FundSubscription, AddConsumer} from "../../script/Interactions.s.sol"; 
`
This is my soln to the `CreateSubscription`

```
function testCreateSubscriptionReturnsSubId() public {
        //Arrange
        CreateSubscription createSubscription = new CreateSubscription();
        uint64 testSubId = createSubscription.createSubscription(
            vrfCoordinator,
            deployerKey
        );
        //Act
        vm.recordLogs();
        VRFCoordinatorV2Mock(vrfCoordinator).createSubscription();
        Vm.Log[] memory entries = vm.getRecordedLogs();
        bytes32 subId = entries[0].topics[1];
        console.log("subId:", uint64(uint160(uint256(subId))));
        console.log("testSubId: ", testSubId);

        //Assert
        assert(testSubId == (uint64(uint160(uint256(subId))) - 1));
        //The mock gives the next subscriptionID! see the working below:
        //function createSubscription() external override returns (uint64 _subId) {
        //s_currentSubId++;
    }
```
The test passes but,
Is that a valid approach?
Also, how do I test for the other two?
Thanks for your time.

**Comment by sharonphiliplima** [comment]
https://github.com/Cyfrin/foundry-smart-contract-lottery-f23/tree/main/script

I want to test `DeployRaffle.s.sol` and `Interactions.s.sol` from the above repo!
![image](https://github.com/Cyfrin/foundry-full-course-f23/assets/69852010/43ddc662-ae7a-4006-87d2-43fcb48d98ab)

**Comment by 0xusmanf** [comment]
I hope [this](https://github.com/usmanfarooq91/decenterlized-lottery-with-foundry/issues/1) helps.

**Comment by RichardPalladino** [comment]
It looks like you've built the test to only work with the mock, so wouldn't the subId be the same every test?  (1, right?)

For the others, check out the corresponding mock contract's functions.  For example, for fund subscription, the contract has a struct called Subscription with an owner and balance that has a corresponding public mapping of s_subscription (which is subId => Subscription).  Just get the balance of your subscription by subId (i.e., your test contract since you're not doing any pranks here).

And if you need help with THAT, there is already a getter function in the mock: getSubscription.  

So, I recommend you read the and understand the mock.

**Reply by sharonphiliplima** [reply]
Thank you, Richard. Is there any other way to pass the tests than relying on the mock?

**Reply by RichardPalladino** [reply]
That just depends on the chain.  You need the mock for a fork.  You shouldn't use the mock for a testnet or production blockchain.

As a side note, and not really applicable to this case, in a regular development process you should always test a staged version of the contract(s) on the production chain before deploying the full production contract.
